File: sdks/node/src/resources/processes.ts
/**
 * `owlette.processes(siteId, machineId)` — scoped process management over
 * `/api/sites/{siteId}/machines/{machineId}/processes[/{processId}[/action]]`.
 *
 * Bound to a (siteId, machineId) tuple and exposed as a factory on the root
 * `Owlette` instance, matching the api's nesting without repeating both ids on
 * every call: `owlette.processes(siteId, machineId).list()`.
 */
import { randomUUID } from 'crypto';
import type { OwletteClient } from '../lib/client';

export type ProcessLaunchMode = 'off' | 'always' | 'scheduled';

export interface ProcessScheduleBlock {
  days: string[];
  ranges: Array<{ start: string; stop: string }>;
}

export interface ProcessSummary {
  processId: string;
  name: string;
  exe_path: string;
  file_path: string;
  cwd: string;
  priority: string;
  visibility: string;
  time_delay: string;
  time_to_init: string;
  relaunch_attempts: string;
  autolaunch: boolean;
  launch_mode: ProcessLaunchMode;
  schedules: ProcessScheduleBlock[] | null;
  schedule: { mode: ProcessLaunchMode; blocks?: ProcessScheduleBlock[] } | null;
  schedulePresetId: string | null;
  status: string;
  pid: number | null;
  responsive: boolean;
  last_updated: string | number | null;
}

export interface ListProcessesResult {
  processes: ProcessSummary[];
  nextPageToken: string | null;
}

export interface CreateProcessOptions {
  name: string;
  exe_path: string;
  file_path?: string;
  cwd?: string;
  priority?: string;
  visibility?: string;
  time_delay?: string;
  time_to_init?: string;
  relaunch_attempts?: string;
  launch_mode?: ProcessLaunchMode;
  schedules?: ProcessScheduleBlock[] | null;
  idempotencyKey?: string;
}

export type UpdateProcessOptions = Partial<
  Omit<CreateProcessOptions, 'idempotencyKey'>
> & {
  idempotencyKey?: string;
};

export interface ScheduleOptions {
  mode: ProcessLaunchMode;
  blocks?: ProcessScheduleBlock[];
  idempotencyKey?: string;
}

/**
 * `restart` response. The server queues an agent command and returns its id at
 * once; poll `owlette.machines.getCommand(siteId, machineId, commandId)`.
 */
export interface RestartProcessResult {
  commandId: string;
  status: string;
}

export class Processes {
  constructor(
    private readonly client: OwletteClient,
    private readonly siteId: string,
    private readonly machineId: string,
  ) {}

  private get base(): string {
    return `/api/sites/${encodeURIComponent(this.siteId)}/machines/${encodeURIComponent(this.machineId)}/processes`;
  }

  async list(): Promise<ListProcessesResult> {
    const res = await this.client.request<{
      ok: true;
      data: ListProcessesResult;
    }>(this.base);
    return res.data.data;
  }

  async get(processId: string): Promise<ProcessSummary> {
    const res = await this.client.request<{ ok: true; data: ProcessSummary }>(
      `${this.base}/${encodeURIComponent(processId)}`,
    );
    return res.data.data;
  }

  async create(opts: CreateProcessOptions): Promise<{ processId: string }> {
    const body: Record<string, unknown> = {
      name: opts.name,
      exe_path: opts.exe_path,
    };
    if (opts.file_path !== undefined) body.file_path = opts.file_path;
    if (opts.cwd !== undefined) body.cwd = opts.cwd;
    if (opts.priority !== undefined) body.priority = opts.priority;
    if (opts.visibility !== undefined) body.visibility = opts.visibility;
    if (opts.time_delay !== undefined) body.time_delay = opts.time_delay;
    if (opts.time_to_init !== undefined) body.time_to_init = opts.time_to_init;
    if (opts.relaunch_attempts !== undefined)
      body.relaunch_attempts = opts.relaunch_attempts;
    if (opts.launch_mode !== undefined) body.launch_mode = opts.launch_mode;
    if (opts.schedules !== undefined) body.schedules = opts.schedules;

    const res = await this.client.request<{
      ok: true;
      data: { processId: string };
    }>(this.base, {
      method: 'POST',
      body,
      idempotencyKey:
        opts.idempotencyKey ?? `sdk-processes-create-${randomUUID()}`,
    });
    return res.data.data;
  }

  async update(
    processId: string,
    opts: UpdateProcessOptions,
  ): Promise<{ processId: string }> {
    const { idempotencyKey, ...patch } = opts;
    const res = await this.client.request<{
      ok: true;
      data: { processId: string };
    }>(`${this.base}/${encodeURIComponent(processId)}`, {
      method: 'PATCH',
      body: patch,
      idempotencyKey: idempotencyKey ?? `sdk-processes-update-${randomUUID()}`,
    });
    return res.data.data;
  }

  async delete(
    processId: string,
  ): Promise<{ processId: string; alreadyDeleted: boolean }> {
    const res = await this.client.request<{
      ok: true;
      data: { processId: string; alreadyDeleted: boolean };
    }>(`${this.base}/${encodeURIComponent(processId)}`, {
      method: 'DELETE',
    });
    return res.data.data;
  }

  async kill(
    processId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<{ ok: true; data: Record<string, unknown> }> {
    return this.controlVerb('kill', processId, opts.idempotencyKey);
  }

  async start(
    processId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<{ ok: true; data: Record<string, unknown> }> {
    return this.controlVerb('start', processId, opts.idempotencyKey);
  }

  async stop(
    processId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<{ ok: true; data: Record<string, unknown> }> {
    return this.controlVerb('stop', processId, opts.idempotencyKey);
  }

  /**
   * Queue a restart. Like `start`/`stop`/`kill`, but returns the command id and
   * status unwrapped. The server stops the process if running then starts it; the
   * agent reports each transition over the command stream.
   */
  async restart(
    processId: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<RestartProcessResult> {
    const res = await this.client.request<{
      ok: true;
      data: RestartProcessResult;
    }>(`${this.base}/${encodeURIComponent(processId)}/restart`, {
      method: 'POST',
      body: {},
      idempotencyKey:
        opts.idempotencyKey ?? `sdk-processes-restart-${randomUUID()}`,
    });
    return res.data.data;
  }

  async schedule(
    processId: string,
    opts: ScheduleOptions,
  ): Promise<{ processId: string; mode: ProcessLaunchMode }> {
    const body: Record<string, unknown> = { mode: opts.mode };
    if (opts.blocks !== undefined) body.blocks = opts.blocks;

    const res = await this.client.request<{
      ok: true;
      data: { processId: string; mode: ProcessLaunchMode };
    }>(`${this.base}/${encodeURIComponent(processId)}/schedule`, {
      method: 'POST',
      body,
      idempotencyKey:
        opts.idempotencyKey ?? `sdk-processes-schedule-${randomUUID()}`,
    });
    return res.data.data;
  }

  private async controlVerb(
    verb: 'kill' | 'start' | 'stop',
    processId: string,
    idempotencyKey: string | undefined,
  ): Promise<{ ok: true; data: Record<string, unknown> }> {
    const res = await this.client.request<{
      ok: true;
      data: Record<string, unknown>;
    }>(`${this.base}/${encodeURIComponent(processId)}/${verb}`, {
      method: 'POST',
      body: {},
      idempotencyKey: idempotencyKey ?? `sdk-processes-${verb}-${randomUUID()}`,
    });
    return res.data;
  }
}

File: sdks/python/roost/resources/processes.py
"""``roost.processes`` — process lifecycle on a single machine.

Drives the wave-2B scoped process api:

  GET    /api/sites/{siteId}/machines/{machineId}/processes
  POST   /api/sites/{siteId}/machines/{machineId}/processes
  GET    /api/sites/{siteId}/machines/{machineId}/processes/{processId}
  PATCH  /api/sites/{siteId}/machines/{machineId}/processes/{processId}
  DELETE /api/sites/{siteId}/machines/{machineId}/processes/{processId}
  POST   /api/sites/{siteId}/machines/{machineId}/processes/{processId}/kill
  POST   /api/sites/{siteId}/machines/{machineId}/processes/{processId}/start
  POST   /api/sites/{siteId}/machines/{machineId}/processes/{processId}/stop
  POST   /api/sites/{siteId}/machines/{machineId}/processes/{processId}/schedule

Construct via the factory ``roost.processes(site_id, machine_id)`` —
each instance is bound to one machine. The factory keeps the noun-verb
shape consistent with the CLI (`owlette process …`) without the caller
having to repeat the site/machine pair on every call.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from roost.client import RoostClient


ScheduleMode = Literal["off", "always", "scheduled"]
ControlVerb = Literal["kill", "start", "stop"]


@dataclass(slots=True)
class ProcessRecord:
    process_id: str
    name: str
    exe_path: str
    cwd: str
    priority: str
    visibility: str
    launch_mode: str
    autolaunch: bool
    status: str
    pid: int | None
    responsive: bool
    schedule: Any | None
    schedules: Any | None
    last_updated: str | int | None


def _parse_process(raw: dict[str, Any]) -> ProcessRecord:
    pid_raw = raw.get("pid")
    pid: int | None = pid_raw if isinstance(pid_raw, int) else None
    return ProcessRecord(
        process_id=str(raw.get("processId", "")),
        name=str(raw.get("name", "")),
        exe_path=str(raw.get("exe_path", "")),
        cwd=str(raw.get("cwd", "")),
        priority=str(raw.get("priority", "")),
        visibility=str(raw.get("visibility", "")),
        launch_mode=str(raw.get("launch_mode", "")),
        autolaunch=bool(raw.get("autolaunch", False)),
        status=str(raw.get("status", "")),
        pid=pid,
        responsive=bool(raw.get("responsive", False)),
        schedule=raw.get("schedule"),
        schedules=raw.get("schedules"),
        last_updated=raw.get("last_updated"),
    )


class Processes:
    """Process management bound to one (site, machine) pair."""

    def __init__(self, client: "RoostClient", site_id: str, machine_id: str) -> None:
        self._client = client
        self._site_id = site_id
        self._machine_id = machine_id

    @property
    def site_id(self) -> str:
        return self._site_id

    @property
    def machine_id(self) -> str:
        return self._machine_id

    def _base(self) -> str:
        return (
            f"/api/sites/{self._site_id}/machines/{self._machine_id}/processes"
        )

    async def list(self) -> list[ProcessRecord]:
        resp = await self._client.request(self._base())
        envelope = resp.data if isinstance(resp.data, dict) else {}
        # Server wraps response in `{ ok, data: { processes, nextPageToken } }`.
        payload = envelope.get("data") if isinstance(envelope.get("data"), dict) else envelope
        if not isinstance(payload, dict):
            return []
        return [
            _parse_process(p)
            for p in (payload.get("processes") or [])
            if isinstance(p, dict)
        ]

    async def get(self, process_id: str) -> ProcessRecord:
        resp = await self._client.request(f"{self._base()}/{process_id}")
        envelope = resp.data if isinstance(resp.data, dict) else {}
        payload = envelope.get("data") if isinstance(envelope.get("data"), dict) else envelope
        return _parse_process(payload if isinstance(payload, dict) else {})

    async def create(
        self,
        *,
        name: str,
        exe_path: str,
        cwd: str | None = None,
        priority: str | None = None,
        visibility: str | None = None,
        launch_mode: ScheduleMode | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"name": name, "exe_path": exe_path}
        if cwd is not None:
            body["cwd"] = cwd
        if priority is not None:
            body["priority"] = priority
        if visibility is not None:
            body["visibility"] = visibility
        if launch_mode is not None:
            body["launch_mode"] = launch_mode

        resp = await self._client.request(
            self._base(),
            method="POST",
            body=body,
            idempotency_key=idempotency_key,
        )
        envelope = resp.data if isinstance(resp.data, dict) else {}
        payload = envelope.get("data") if isinstance(envelope.get("data"), dict) else envelope
        return payload if isinstance(payload, dict) else {}

    async def update(
        self,
        process_id: str,
        *,
        name: str | None = None,
        exe_path: str | None = None,
        cwd: str | None = None,
        priority: str | None = None,
        visibility: str | None = None,
        launch_mode: ScheduleMode | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if exe_path is not None:
            body["exe_path"] = exe_path
        if cwd is not None:
            body["cwd"] = cwd
        if priority is not None:
            body["priority"] = priority
        if visibility is not None:
            body["visibility"] = visibility
        if launch_mode is not None:
            body["launch_mode"] = launch_mode
        if not body:
            msg = "processes.update: at least one field must be provided"
            raise ValueError(msg)

        resp = await self._client.request(
            f"{self._base()}/{process_id}",
            method="PATCH",
            body=body,
            idempotency_key=idempotency_key,
        )
        envelope = resp.data if isinstance(resp.data, dict) else {}
        payload = envelope.get("data") if isinstance(envelope.get("data"), dict) else envelope
        return payload if isinstance(payload, dict) else {}

    async def delete(self, process_id: str) -> dict[str, Any]:
        resp = await self._client.request(
            f"{self._base()}/{process_id}",
            method="DELETE",
        )
        envelope = resp.data if isinstance(resp.data, dict) else {}
        payload = envelope.get("data") if isinstance(envelope.get("data"), dict) else envelope
        return payload if isinstance(payload, dict) else {}

    async def kill(
        self, process_id: str, *, idempotency_key: str | None = None
    ) -> dict[str, Any]:
        return await self._control(process_id, "kill", idempotency_key)

    async def start(
        self, process_id: str, *, idempotency_key: str | None = None
    ) -> dict[str, Any]:
        return await self._control(process_id, "start", idempotency_key)

    async def stop(
        self, process_id: str, *, idempotency_key: str | None = None
    ) -> dict[str, Any]:
        return await self._control(process_id, "stop", idempotency_key)

    async def _control(
        self,
        process_id: str,
        verb: ControlVerb,
        idempotency_key: str | None,
    ) -> dict[str, Any]:
        resp = await self._client.request(
            f"{self._base()}/{process_id}/{verb}",
            method="POST",
            body={},
            idempotency_key=idempotency_key,
        )
        envelope = resp.data if isinstance(resp.data, dict) else {}
        payload = envelope.get("data") if isinstance(envelope.get("data"), dict) else envelope
        return payload if isinstance(payload, dict) else {}

    async def schedule(
        self,
        process_id: str,
        *,
        mode: ScheduleMode,
        blocks: Sequence[dict[str, Any]] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        if mode == "scheduled" and (blocks is None or len(blocks) == 0):
            msg = "processes.schedule: `blocks` is required when mode='scheduled'"
            raise ValueError(msg)
        body: dict[str, Any] = {"mode": mode}
        if blocks is not None:
            body["blocks"] = list(blocks)
        resp = await self._client.request(
            f"{self._base()}/{process_id}/schedule",
            method="POST",
            body=body,
            idempotency_key=idempotency_key,
        )
        envelope = resp.data if isinstance(resp.data, dict) else {}
        payload = envelope.get("data") if isinstance(envelope.get("data"), dict) else envelope
        return payload if isinstance(payload, dict) else {}


__all__ = ["ControlVerb", "ProcessRecord", "Processes", "ScheduleMode"]
